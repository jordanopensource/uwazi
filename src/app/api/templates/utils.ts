import uuid from 'node-uuid';
import { ObjectId } from 'mongodb';
import { differenceBy, intersectionBy } from 'lodash';

import settings from 'api/settings/settings';
import { files } from 'api/files';
import propertiesHelper from 'shared/commonProperties';
import { objectIndex } from 'shared/data_utils/objectIndex';
import { safeName as sharedSafeName } from 'shared/propertyNames';
import { ensure } from 'shared/tsUtils';
import { ExtractedMetadataSchema, PropertySchema } from 'shared/types/commonTypes';
import { TemplateSchema } from 'shared/types/templateType';
import { ThesaurusSchema, ThesaurusValueSchema } from 'shared/types/thesaurusType';
import model from './templatesModel';

const safeName = sharedSafeName;

const getInheritedProps = async (templates: TemplateSchema[]) => {
  const properties: PropertySchema[] = propertiesHelper
    .allUniqueProperties(templates)
    .filter((p: PropertySchema) => p.inherit?.property);

  return (
    await model.db.aggregate([
      {
        $match: {
          'properties._id': {
            $in: properties.map(p => new ObjectId(p.inherit?.property)),
          },
        },
      },
      {
        $project: {
          properties: {
            $filter: {
              input: '$properties',
              as: 'property',
              cond: {
                $or: properties.map(p => ({
                  $eq: ['$$property._id', new ObjectId(p.inherit?.property)],
                })),
              },
            },
          },
          _id: 0,
        },
      },
      { $unwind: '$properties' },
      { $replaceRoot: { newRoot: '$properties' } },
    ])
  ).reduce((indexed, prop) => ({ ...indexed, [prop._id.toString()]: prop }), {});
};

const denormalizeInheritedProperties = async (template: TemplateSchema) => {
  if (template.synced) return template.properties;

  const inheritedProperties: { [k: string]: PropertySchema } = await getInheritedProps([template]);

  return template.properties?.map(prop => {
    if (!prop.inherit?.property) {
      // eslint-disable-next-line no-param-reassign
      delete prop.inherit;
      return prop;
    }

    const { type } = inheritedProperties[prop.inherit.property];
    // eslint-disable-next-line no-param-reassign
    prop.inherit.type = type;
    return prop;
  });
};

const generateName = (property: PropertySchema, newNameGeneration: boolean) => {
  const name = property.label ? safeName(property.label, newNameGeneration) : property.name;
  return property.type === 'geolocation' || property.type === 'nested'
    ? `${name}_${property.type}`
    : name;
};

const generateNames = async (properties: PropertySchema[]) => {
  const { newNameGeneration = false } = await settings.get();
  return properties.map(property => ({
    ...property,
    name: generateName(property, newNameGeneration),
  }));
};

const newThesauriId = () => uuid.v4();

function generateIds(properties: ThesaurusSchema[] = []) {
  return properties.map(property => ({
    ...property,
    id: property.id || newThesauriId(),
  }));
}
interface PropertyOrThesaurusSchema
  extends Partial<PropertySchema>,
    Partial<ThesaurusValueSchema> {}

const flattenProperties = (properties: PropertyOrThesaurusSchema[]) =>
  properties.reduce<PropertyOrThesaurusSchema[]>((flatProps, p) => {
    if (p.values) {
      return [...flatProps, ...p.values, p];
    }

    return [...flatProps, p];
  }, []);

function getUpdatedIds(
  {
    prop,
    filterBy,
  }: {
    prop: 'name' | 'label';
    filterBy: 'id' | '_id';
  },
  oldProperties: PropertyOrThesaurusSchema[] = [],
  newProperties: PropertyOrThesaurusSchema[] = []
) {
  const indexOf = (p: PropertyOrThesaurusSchema) => p[filterBy]!.toString();
  const idsWithNewName: { [k: string]: string } = {};
  const flatOld = flattenProperties(oldProperties);
  const flatNew = flattenProperties(newProperties);
  const newByIndex = objectIndex(flatNew, indexOf, p => p);
  flatOld.forEach(property => {
    const newProperty = newByIndex[indexOf(property)];
    const oldValue = property[prop];
    const newValue = newProperty?.[prop];
    if (newProperty && typeof newValue === 'string' && newValue !== oldValue) {
      idsWithNewName[indexOf(property)] = newValue;
    }
  });
  return idsWithNewName;
}

function getUpdatedNames(
  {
    prop,
    filterBy,
  }: {
    prop: 'name' | 'label';
    filterBy: 'id' | '_id';
  },
  oldProperties: PropertyOrThesaurusSchema[] = [],
  newProperties: PropertyOrThesaurusSchema[] = []
) {
  const indexOf = (p: PropertyOrThesaurusSchema) => p[filterBy]?.toString() || '';

  const propertiesWithNewName: { [k: string]: string } = {};
  const deletedProperties: Set<string> = new Set();

  const flatOld = flattenProperties(oldProperties);
  const preExistingLabels = new Set(flatOld.map(property => property[prop]));
  const flatNew = flattenProperties(newProperties);
  const newByIndex = objectIndex(flatNew, indexOf, p => p);
  const allNewLabels = new Set(flatNew.map(property => property[prop]));

  const labelAlreadyExisted = (p: PropertyOrThesaurusSchema) => preExistingLabels.has(p[prop]);
  const labelStillInUse = (p: PropertyOrThesaurusSchema) => allNewLabels.has(p[prop]);
  const sameProp = (previous: PropertyOrThesaurusSchema, current: PropertyOrThesaurusSchema) =>
    previous[prop] === current[prop];
  const compareProps = (
    previous: PropertyOrThesaurusSchema,
    current: PropertyOrThesaurusSchema
  ) => {
    if (!current || sameProp(previous, current) || labelAlreadyExisted(current)) {
      if (!labelStillInUse(previous)) deletedProperties.add(previous[prop]!);
      return { key: undefined, value: undefined };
    }
    if (labelStillInUse(previous)) {
      return { key: current[prop], value: current[prop] };
    }
    return { key: previous[prop], value: current[prop] };
  };

  flatOld.forEach(property => {
    const newProperty = newByIndex[indexOf(property)];
    const { key, value } = compareProps(property, newProperty);
    if (key && value) {
      propertiesWithNewName[key] = value;
    }
  });

  return { update: propertiesWithNewName, delete: Array.from(deletedProperties) };
}

function getDeletedProperties(
  oldProperties: PropertyOrThesaurusSchema[] = [],
  newProperties: PropertyOrThesaurusSchema[] = [],
  filterBy: 'id' | '_id' = '_id',
  prop: 'name' | 'label' | 'id' = 'name'
) {
  const flatOld = flattenProperties(oldProperties);
  const flatNew = flattenProperties(newProperties);
  const newIds = new Set(flatNew.map(property => property[filterBy]?.toString()));
  const newLabels = new Set(flatNew.map(property => property[prop]?.toString()));
  const filteredById = flatOld
    .filter(
      property => !newIds.has(property[filterBy]?.toString()) && !newLabels.has(property[prop])
    )
    .filter((property): property is { [k: string]: string } => typeof property[prop] === 'string')
    .map(property => property[prop]);
  return filteredById;
}

function getRenamedTitle(
  oldCommonProperties: PropertySchema[],
  newCommonProperties: PropertySchema[]
) {
  const oldTitle = ensure<PropertySchema>(oldCommonProperties.find(p => p.name === 'title'));
  const newTitle = ensure<PropertySchema>(newCommonProperties.find(p => p.name === 'title'));
  return oldTitle.label !== newTitle.label ? [oldTitle.label] : [];
}

const propertyUpdater = async (
  modifiedProperties: PropertySchema[] = [],
  updateFunction: (
    array: ExtractedMetadataSchema[],
    property: PropertySchema
  ) => ExtractedMetadataSchema[] = () => []
) =>
  modifiedProperties.reduce(async (previousPromise: Promise<void>, property) => {
    await previousPromise;
    const affectedFiles = await files.get({
      'extractedMetadata.propertyID': property._id?.toString(),
    });

    await affectedFiles.reduce(async (prevPromise: Promise<void>, file) => {
      await prevPromise;
      await files.save({
        ...file,
        extractedMetadata: updateFunction(file.extractedMetadata || [], property),
      });
    }, Promise.resolve());
  }, Promise.resolve());

const updateExtractedMetadataProperties = async (
  oldProperties: PropertySchema[] = [],
  newProperties: PropertySchema[] = []
) => {
  const currentProperties = oldProperties.map(property => ({
    ...property,
    _id: property._id?.toString(),
  }));

  const differentProperties = differenceBy(newProperties, currentProperties, 'name').filter(
    property => ['text', 'markdown', 'numeric', 'date'].includes(property.type)
  );

  const renamedProperties = intersectionBy(differentProperties, currentProperties, '_id');

  const removedProperties = differenceBy(currentProperties, newProperties, '_id').filter(property =>
    ['text', 'markdown', 'numeric', 'date'].includes(property.type)
  );

  if (removedProperties.length > 0) {
    await propertyUpdater(removedProperties, (metadata, property) =>
      metadata.filter(data => data.propertyID !== property._id?.toString())
    );
  }

  if (renamedProperties.length > 0) {
    await propertyUpdater(renamedProperties, (metadata, property) =>
      metadata.map(data => {
        if (data.propertyID === property._id?.toString()) {
          return { ...data, name: property.name };
        }
        return data;
      })
    );
  }

  return null;
};

export type { PropertyOrThesaurusSchema };
export {
  newThesauriId,
  safeName,
  denormalizeInheritedProperties,
  flattenProperties,
  generateIds,
  getUpdatedIds,
  getUpdatedNames,
  generateNames,
  getDeletedProperties,
  getRenamedTitle,
  updateExtractedMetadataProperties,
};
