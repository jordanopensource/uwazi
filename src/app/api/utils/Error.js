/**
 * @deprecated
 */
export default function (message, code = 500, logLevel = 'debug') {
  if (code === 500) {
    logLevel = 'error';
  }

  if (message.ajv) {
    return { message: message.message, ...message, code, logLevel };
  }

  if (message instanceof Error) {
    return { message: message.message, stack: message.stack, code, original: message, logLevel };
  }

  return { message, code, logLevel };
}
