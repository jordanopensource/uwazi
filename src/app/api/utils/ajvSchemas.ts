const ObjectIdAsString = {
  type: 'string',
  pattern: '^[0-9a-fA-F]{24}$',
};

export { ObjectIdAsString };
