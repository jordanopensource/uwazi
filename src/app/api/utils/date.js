import moment from 'moment';
// @ts-ignore
import parser from 'any-date-parser';
import * as stopword from 'stopword';

export default {
  currentUTC() {
    return moment.utc().toDate().getTime();
  },

  descriptionToTimestamp(date) {
    if (date === 'last-day-last-month') {
      return moment.utc().subtract(1, 'months').endOf('month').unix();
    }

    if (date === 'first-day-last-month') {
      return moment.utc().subtract(1, 'months').startOf('month').unix();
    }

    return date;
  },

  addYearsToCurrentDate(yearsToAdd) {
    const newDate = new Date();
    newDate.setHours(0, 0, 0, 0);
    newDate.setFullYear(newDate.getFullYear() + yearsToAdd);
    return newDate;
  },

  dateToSeconds(value, locale) {
    let parsedValue = value.normalize('NFD').replace(/[\u0300-\u036f]/g, '');

    const { removeStopwords, ...languages } = stopword;

    if (locale) {
      parsedValue = removeStopwords(parsedValue.split(' '), languages[locale]).join(' ');
    }

    let getDate = parser.fromString(parsedValue, locale);
    if (getDate.invalid) {
      getDate = Date.parse(`${parsedValue} GMT`);
    }
    const formattedDate = getDate / 1000;
    return formattedDate;
  },
};
