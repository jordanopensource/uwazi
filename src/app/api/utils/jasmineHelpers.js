/** @format */

const assessStatus = routeResult => {
  let status;
  const req = {};
  const next = () => {};
  const res = {
    status: code => {
      status = code;
    },
    json: () => {},
  };

  routeResult.middlewares[0](req, res, next);

  return status;
};

const assessAuthorized = (routeResult, roles = []) => {
  const nextCalledFor = [];

  roles.forEach(role => {
    const req = {
      user: { role },
      get: key => (key === 'X-Requested-With' ? 'XMLHttpRequest' : ''),
    };
    const res = { status: () => {}, json: () => {} };
    const next = () => {
      nextCalledFor.push(role);
    };
    routeResult.middlewares[0](req, res, next);
  });

  return roles.join(',') === nextCalledFor.join(',');
};

const conformValidationErrors = (routeCanFail, expected) =>
  routeCanFail
    ? `Route is not correctly authorized for ['${expected.join("', '")}']`
    : 'Route is not authorized ! (Auth middleware should be the first one)';

const conformMessage = (result, routeCanFail, expected) =>
  result.pass ? 'route is authorized' : conformValidationErrors(routeCanFail, expected);

const matchers = {
  toNeedAuthorization(routeResult, expected) {
    const routeCanFail = assessStatus(routeResult) === 401;
    const routeValidatesExpected = expected ? assessAuthorized(routeResult, expected) : true;
    const result = { pass: routeCanFail && routeValidatesExpected };
    result.message = () => conformMessage(result, routeCanFail, expected);
    return result;
  },
};

expect.extend(matchers);

export function catchErrors(done) {
  return error => {
    if (error instanceof Error) {
      return done.fail(error);
    }
    return done.fail(JSON.stringify(error));
  };
}
