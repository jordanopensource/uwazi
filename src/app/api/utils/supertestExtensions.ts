import { Response } from 'supertest';
import Test from 'supertest/lib/test';

function extractStatusDebugInfo(res: Response): string {
  try {
    return JSON.stringify(JSON.parse(res.text), null, 2);
  } catch (e) {
    return res.text;
  }
}

export function extendSupertest() {
  //@ts-ignore
  const { _assertStatus } = Test.prototype;

  //@ts-ignore
  Test.prototype._assertStatus = function extendedAssertStatus(status: number, res: Response) {
    const err: Error = _assertStatus(status, res);
    if (err) {
      err.message += `\n ${extractStatusDebugInfo(res)}`;
      delete err.stack;
    }
    return err;
  };
}
