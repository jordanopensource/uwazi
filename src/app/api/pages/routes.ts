import { Application, Request } from 'express';

import { validation } from 'api/utils';
import needsAuthorization from '../auth/authMiddleware';
import pages from './pages';

export default (app: Application) => {
  app.post('/api/pages', needsAuthorization(['admin']), (req, res, next) => {
    pages
      .save(req.body, req.user, req.language)
      .then(response => res.json(response))
      .catch(next);
  });

  app.get(
    '/api/pages',
    validation.validateRequest({
      type: 'object',
      properties: {
        query: {
          type: 'object',
          properties: {
            sharedId: {
              type: 'string',
            },
          },
        },
      },
      required: ['query'],
    }),
    (req, res, next) => {
      pages
        .get({ ...req.query, language: req.language })
        .then(res.json.bind(res))
        .catch(next);
    }
  );

  app.get(
    '/api/page',
    validation.validateRequest({
      type: 'object',
      properties: {
        query: {
          type: 'object',
          properties: {
            sharedId: {
              type: 'string',
            },
            slug: {
              type: 'string',
            },
          },
        },
      },
      required: ['query'],
    }),
    (req: Request<{}, {}, {}, { sharedId: string }>, res, next) => {
      pages.getById(req.query.sharedId, req.language).then(res.json.bind(res)).catch(next);
    }
  );

  app.delete(
    '/api/pages',
    needsAuthorization(),
    validation.validateRequest({
      type: 'object',
      properties: {
        query: {
          type: 'object',
          properties: {
            sharedId: {
              type: 'string',
            },
          },
        },
      },
      required: ['query'],
    }),
    (req: Request<{}, {}, {}, { sharedId: string }>, res, next) => {
      pages
        .delete(req.query.sharedId)
        .then(response => res.json(response))
        .catch(next);
    }
  );
};
