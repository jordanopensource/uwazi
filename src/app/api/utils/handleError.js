import Ajv from 'ajv';
import { UnauthorizedError } from 'api/authorization.v2/errors/UnauthorizedError';
import { ValidationError } from 'api/common.v2/validation/ValidationError';
import { FileNotFound } from 'api/files/FileNotFound';
import { S3TimeoutError } from 'api/files/S3Storage';
import { legacyLogger } from 'api/log';
import { appContext } from 'api/utils/AppContext';
import { createError } from 'api/utils/index';
import util from 'node:util';

const ajvPrettifier = error => {
  const errorMessage = [error.message];
  if (error.validations && error.validations.length) {
    error.validations.forEach(oneError => {
      errorMessage.push(`${oneError.instancePath}: ${oneError.message}`);
    });
  }
  return errorMessage.join('\n');
};

const fallbackPrettifier = (error, req) => {
  const url = req.originalUrl ? `\nurl: ${req.originalUrl}` : '';
  const body =
    req.body && Object.keys(req.body).length
      ? `\nbody: ${JSON.stringify(req.body, null, ' ')}`
      : '';
  const query =
    req.query && Object.keys(req.query).length
      ? `\nquery: ${JSON.stringify(req.query, null, ' ')}`
      : '';
  const errorString = `\n${error.message || JSON.stringify(error.json)}`;

  let errorMessage = `${url}${body}${query}${errorString}`;

  //if the resulting message is empty, or meaningless combination of characters ('{}')
  if (errorMessage.match(/^[{}\s]*$/g)) {
    errorMessage = JSON.stringify(error, null, 2);
  }

  return errorMessage;
};

const appendOriginalError = (message, originalError) =>
  `${message}\noriginal error: ${JSON.stringify(originalError, null, ' ')}`;

const obfuscateCredentials = req => {
  const obfuscated = req;
  if (req.body && req.body.password) {
    obfuscated.body.password = '########';
  }

  if (req.body && req.body.username) {
    obfuscated.body.username = '########';
  }

  return obfuscated;
};

// eslint-disable-next-line max-statements
const prettifyError = (error, { req = {}, uncaught = false } = {}) => {
  let result = error;

  if (error instanceof Error) {
    result = { code: 500, message: util.inspect(error), logLevel: 'error' };
  }

  if (error instanceof S3TimeoutError) {
    result = { code: 408, message: util.inspect(error), logLevel: 'debug' };
  }

  if (error instanceof Ajv.ValidationError) {
    result = { code: 422, message: error.message, validations: error.errors, logLevel: 'debug' };
  }

  if (error.name === 'ValidationError') {
    result = {
      code: 422,
      message: error.message,
      validations: error.properties,
      logLevel: 'debug',
    };
  }

  if (error instanceof ValidationError) {
    result = { code: 422, message: error.message, validations: error.errors, logLevel: 'debug' };
  }

  if (error instanceof UnauthorizedError) {
    result = { code: 401, message: error.message, logLevel: 'debug' };
  }

  if (error instanceof FileNotFound) {
    result = { code: 404, message: error.message, logLevel: 'debug' };
  }

  if (error.name === 'MongoError') {
    result.code = 500;
    result.logLevel = 'error';
  }

  if (error.message && error.message.match(/Cast to ObjectId failed for value/)) {
    result.code = 400;
    result.logLevel = 'debug';
  }

  if (error.message && error.message.match(/rison decoder error/)) {
    result.code = 400;
    result.logLevel = 'debug';
  }

  if (uncaught) {
    result.message = `uncaught exception or unhandled rejection, Node process finished !!\n ${result.message}`;
    result.logLevel = 'error';
    result.code = 500;
  }

  const obfuscatedRequest = obfuscateCredentials(req);
  result.prettyMessage = error.ajv
    ? ajvPrettifier(result)
    : fallbackPrettifier(result, obfuscatedRequest);

  return result;
};

const getErrorMessage = (data, error) => {
  const originalError = data.original || error;
  const prettyMessage = data.requestId
    ? `requestId: ${data.requestId} ${data.prettyMessage}`
    : data.prettyMessage;

  if (originalError instanceof Error) {
    const extendedError = appendOriginalError(prettyMessage, originalError);
    return data.tenantError
      ? `${extendedError}\n[Tenant error] ${data.tenantError.message}`
      : extendedError;
  }

  return prettyMessage;
};

const sendLog = (data, error, errorOptions) => {
  const messageToLog = getErrorMessage(data, error);
  if (data.logLevel === 'debug') {
    legacyLogger.debug(messageToLog, errorOptions);
    return;
  }

  legacyLogger.error(messageToLog, errorOptions);
};

function setRequestId(result) {
  try {
    return { ...result, requestId: appContext.get('requestId') };
  } catch (err) {
    return { ...result, tenantError: err };
  }
}

const simplifyError = (result, original) => {
  const processedError = { ...result };
  delete processedError.original;

  if (original instanceof Error && processedError.code === 500 && original.name !== 'MongoError') {
    processedError.prettyMessage = original.message;
    processedError.error = original.message;
    delete processedError.message;
  } else {
    processedError.prettyMessage = processedError.prettyMessage || original.message;
  }

  return processedError;
};

const handleError = (_error, { req = {}, uncaught = false, useContext = true } = {}) => {
  const errorData = typeof _error === 'string' ? createError(_error, 500) : _error;

  const error = errorData || new Error('Unexpected error has occurred');
  let result = prettifyError(error, { req, uncaught });

  if (useContext) {
    result = setRequestId(result);
  }

  sendLog(result, error, {});

  result = simplifyError(result, error);

  return result;
};

export { handleError, prettifyError };
