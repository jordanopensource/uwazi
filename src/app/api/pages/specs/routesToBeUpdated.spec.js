import 'api/utils/jasmineHelpers';
import instrumentRoutes from '../../utils/instrumentRoutes';
import pages from '../pages';
import pagesRoutes from '../routes';

describe('Pages Routes (to be updated)', () => {
  let routes;

  beforeEach(() => {
    routes = instrumentRoutes(pagesRoutes);
  });

  describe('POST', () => {
    let req;
    beforeEach(() => {
      req = {
        body: { title: 'Batman begins' },
        user: { username: 'admin' },
        language: 'lang',
      };
    });

    it('should need authorization', () => {
      jest.spyOn(pages, 'save').mockImplementation(async () => Promise.resolve('pages'));
      expect(routes.post('/api/pages', req)).toNeedAuthorization();
    });

    it('should create a new document with current user', async () => {
      jest.spyOn(pages, 'save').mockImplementation(async () => Promise.resolve('document'));
      const document = await routes.post('/api/pages', req);
      expect(document).toBe('document');
      expect(pages.save).toHaveBeenCalledWith(req.body, req.user, 'lang');
    });
  });

  describe('/api/pages', () => {
    it('should ask pages model for the page in the current locale', async () => {
      const req = {
        query: { sharedId: '123' },
        language: 'es',
      };
      jest.spyOn(pages, 'get').mockImplementation(async () => Promise.resolve('page'));
      const response = await routes.get('/api/pages', req);
      expect(pages.get).toHaveBeenCalledWith({ sharedId: '123', language: 'es' });
      expect(response).toBe('page');
    });
  });

  describe('/api/page', () => {
    it('should ask pages model for the page in the current locale', async () => {
      const req = {
        query: { sharedId: '123' },
        language: 'es',
      };
      jest.spyOn(pages, 'getById').mockImplementation(async () => Promise.resolve('page'));
      const response = await routes.get('/api/page', req);
      expect(pages.getById).toHaveBeenCalledWith('123', 'es');
      expect(response).toBe('page');
    });
  });

  describe('GET', () => {
    describe('/api/pages', () => {
      it('should have a validation schema', () => {
        expect(routes.get.validation('/api/pages')).toMatchSnapshot();
      });
    });
  });

  describe('DELETE', () => {
    beforeEach(() => {
      jest.spyOn(pages, 'delete').mockImplementation(async () => Promise.resolve({ json: 'ok' }));
    });

    it('should have a validation schema', () => {
      expect(routes.delete.validation('/api/pages')).toMatchSnapshot();
    });

    it('should use pages to delete it', async () => {
      const req = { query: { _id: 123, _rev: 456, sharedId: '456' } };
      await routes.delete('/api/pages', req);
      expect(pages.delete).toHaveBeenCalledWith(req.query.sharedId);
    });
  });
});
