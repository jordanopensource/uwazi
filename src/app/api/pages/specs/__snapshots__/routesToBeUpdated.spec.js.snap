// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages Routes (to be updated) DELETE should have a validation schema 1`] = `
Object {
  "properties": Object {
    "query": Object {
      "properties": Object {
        "sharedId": Object {
          "type": "string",
        },
      },
      "type": "object",
    },
  },
  "required": Array [
    "query",
  ],
  "type": "object",
}
`;

exports[`Pages Routes (to be updated) GET /api/pages should have a validation schema 1`] = `
Object {
  "properties": Object {
    "query": Object {
      "properties": Object {
        "sharedId": Object {
          "type": "string",
        },
      },
      "type": "object",
    },
  },
  "required": Array [
    "query",
  ],
  "type": "object",
}
`;
